pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -1062,17 +1069,26 @@ private void makeConsumerWrapper(String reason) {
   // This should be done during commit? We may not always commit when we build a segment....
   // TODO Call this method when we are loading the segment, which we do from table datamanager afaik
   private void updateCurrentDocumentCountMetrics() {
-    int currentRawDocs = _realtimeSegment.getNumDocsIndexed();
-    _serverMetrics.addValueToTableGauge(_tableName, ServerGauge.DOCUMENT_COUNT, (currentRawDocs - _lastUpdatedRawDocuments
-        .get()));
-    _lastUpdatedRawDocuments.set(currentRawDocs);
+
+    // When updating of metrics is enabled, numRowsIndexed can be <= numRowsConsumed. This is because when in this
+    // case when a new row with existing dimension combination comes in, we find the existing row and update metrics.
+
+    // Number of rows indexed should be used for DOCUMENT_COUNT metric, and also for segment flush. Whereas,
+    // Number of rows consumed should be used for consumption metric.
+    long rowsIndexed = _numRowsIndexed - _lastUpdatedRawDocuments.get();
```

Comment:
So, lastUpdatedRawDocuments now has the number of rows indexed (instead of raw docs). You may want to rename it as such.  Perhaps this var can be a volatile int instead of AtomicLong ... not sure why it is that way.

Reply:
Ran out of direct memory while trying to allocate 214721752 bytes